airflow-core/src/airflow/api_fastapi/core_api/routes/ui/grid.py

```diff
@@ -257,8 +268,254 @@ def grid_data(
+        return nodes
+
+    serdags = session.scalars(
+        select(SerializedDagModel).where(
+            SerializedDagModel.dag_version_id.in_(
+                select(TaskInstance.dag_version_id)
+                .join(TaskInstance.dag_run)
+                .where(DagRun.id.in_(run_ids), SerializedDagModel.id != current_serdag.id)
+            )
+        )
+    )
+    merged_nodes: list[GridNodeResponse] = []
+    dags = [latest_dag]
+    for serdag in serdags:
+        if serdag:
+            dags.append(serdag.dag)
+    for dag in dags:
+        nodes = [task_group_to_dict_grid(x) for x in task_group_sort(dag.task_group)]
+        _merge_node_dicts(merged_nodes, nodes)
+
+    return merged_nodes
 
-    flat_tis = itertools.chain.from_iterable(tis_by_run_id.values())
-    structure = get_combined_structure(task_instances=flat_tis, session=session)
 
-    return GridResponse(dag_runs=grid_dag_runs, structure=structure)
+@grid_router.get(
+    "/runs/{dag_id}",
+    responses=create_openapi_http_exception_doc([status.HTTP_400_BAD_REQUEST, status.HTTP_404_NOT_FOUND]),
+    dependencies=[
+        Depends(requires_access_dag(method="GET", access_entity=DagAccessEntity.TASK_INSTANCE)),
+        Depends(requires_access_dag(method="GET", access_entity=DagAccessEntity.RUN)),
+    ],
+    response_model_exclude_none=True,
+)
+def get_grid_runs(
+    dag_id: str,
+    session: SessionDep,
+    offset: QueryOffset,
+    limit: QueryLimit,
+    order_by: Annotated[
+        SortParam,
+        Depends(
+            SortParam(
+                [
+                    "run_after",
+                    "logical_date",
+                    "start_date",
+                    "end_date",
+                ],
+                DagRun,
+            ).dynamic_depends()
+        ),
+    ],
+    run_after: Annotated[RangeFilter, Depends(datetime_range_filter_factory("run_after", DagRun))],
+) -> list[GridRunsResponse]:
+    """Return unified dag structure for grid view."""
+    current_serdag = session.scalar(
+        select(SerializedDagModel)
+        .where(
+            SerializedDagModel.dag_id == dag_id,
+        )
+        .order_by(SerializedDagModel.id.desc())
+        .limit(1)
+    )
+    if not current_serdag:
+        raise HTTPException(status.HTTP_404_NOT_FOUND, f"Dag with id {dag_id} was not found")
+    # todo: let's load serdag once maybe?
+    # todo: runs for dag should be faster
+    latest_dag = current_serdag.dag
+
+    # Retrieve, sort the previous DAG Runs
+    base_query = select(
+        DagRun.dag_id,
+        DagRun.run_id,
+        DagRun.queued_at,
+        DagRun.start_date,
+        DagRun.end_date,
+        DagRun.run_after,
+        DagRun.state,
+        DagRun.run_type,
+    ).where(DagRun.dag_id == latest_dag.dag_id)
+
+    # This comparison is to falls to DAG timetable when no order_by is provided
+    if order_by.value == order_by.get_primary_key_string():
```

Comment:
Does it make sense to defer looking up the latest dag serdag at all, unless we need the default order_by? Probably worth a follow up at this point though.

Reply:
good eye